hsm/hsm.go

```diff
@@ -0,0 +1,291 @@
+				fmt.Printf("Private Key\n")
+			default:
+				fmt.Printf("%d\n", a.Value[0])
+			}
+		case pkcs11.CKA_ID:
+			fmt.Printf("ID: %x\n", a.Value)
+		default:
+			fmt.Printf("0x%04X: %s (len %d)\n", a.Type, hex.EncodeToString(a.Value), len(a.Value))
+		}
+	}
+
+	return nil
+}
+
+func (h *HSM) FindObjectsByLabel(label string, max int) error {
+	// Set the search parameters
+	t := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+	}
+
+	if err := h.api.FindObjectsInit(h.session, t); err != nil {
+		return ErrQueryFailure
+	}
+
+	// Get the first object that matches t
+	o, _, err := h.api.FindObjects(h.session, max)
+	if err != nil {
+		return ErrQueryFailure
+	}
+
+	// Clean up
+	if err = h.api.FindObjectsFinal(h.session); err != nil {
+		return ErrQueryFailure
+	}
+
+	// Display matching object
+	fmt.Printf("Found %d objects with label \"%s\"\n", len(o), label)
+	for i, obj := range o {
+		fmt.Printf("  Object %d:\n", i)
+		if err := h.displayObject(obj); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *HSM) FindObjectsByUUID(u uuid.UUID, max int) error {
+	// Set the search parameters
+	t := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_ID, []byte(u[0:])),
+	}
+
+	if err := h.api.FindObjectsInit(h.session, t); err != nil {
+		return ErrQueryFailure
+	}
+
+	// Get the first object that matches t
+	o, _, err := h.api.FindObjects(h.session, max)
+	if err != nil {
+		return ErrQueryFailure
+	}
+
+	// Clean up
+	if err = h.api.FindObjectsFinal(h.session); err != nil {
+		return ErrQueryFailure
+	}
+
+	// Display matching object
+	fmt.Printf("Found %d objects with UUID \"%s\"\n", len(o), u.String())
+	for i, obj := range o {
+		fmt.Printf("  Object %d:\n", i)
+		if err := h.displayObject(obj); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Determined with:
+// $ openssl ecparam --name prime256v1 -outform DER | hexdump -C
+var p256Oid []byte = []byte{0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07}
+
+func (h *HSM) CreateRootCAKey() error {
+	// Private key settins
+	// t := []*pkcs11.Attribute{
+	// 	pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_EC),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_PRIVATE, true),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_LABEL, CA_KEY_LABEL),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_ID, CA_KEY_UUID),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+	// 	pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, false),
+	// 	// TODO: Add other required attributes
+	// }
+
+	pubAtts := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, p256Oid),
+	}
+	privAtts := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+	}
```

Comment:
Hopefully this defaults to something secure with the `CKM_EC_KEY_PAIR_GEN` mechanism, but do we need to explicitly set the following attributes at a minimum?

```go
pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, false),
```

Not sure if `GenerateKeyPair` automatically sets `CKA_PRIVATE` to `true`either, but easy to verify once it's created, I guess.